apps/emqx/src/emqx_channel.erl

```diff
@@ -2227,15 +2237,152 @@ session_disconnect(_ClientInfo, _ConnInfo, undefined) ->
 
 %%--------------------------------------------------------------------
 %% Maybe Publish will msg
-
-maybe_publish_will_msg(Channel = #channel{will_msg = undefined}) ->
+%% @doc May publish will message [MQTT-3.1.2-8]
+%%   When willmsg presents the decision whether or when to publish the Will Message are effected by
+%%   the followings:
+%%   - connecion state
+%%   - If it is MQTT normal disconnection (RC: 0)
+%%   - If it is MQTT normal disconnection (RC: 4)
+%%   - will delay interval (MQTT 5.0 only)
+%%   - session expire Session Expiry (MQTT 5.0 only)
+%%   - EMQX operations on the client
+%% @NOTE:
+%%    Connection close with session expiry interval = 0 means session close.
+%% @NOTE:
+%%    The caller does not need to take care of the case when process terminates while will_msg is published
+%%    as it is designed by the spec.
+-spec maybe_publish_will_msg(Reason, channel()) -> channel() when
+    Reason ::
+        %% Connection is terminating because session is taken over by another process.
+        takenover
+        %% Connection is terminating because of EMQX mgmt operation, also delete the session.
+        | kicked
+        %% Connection is terminating because session is taken over by another process.
+        | discarded
+        %% Connection is terminating because session is expired
+        | expired
+        %% Connection is terminating because of socket close/error
+        | sock_closed
+        %% Connection is terminating with Reasons
+        | {shutdown, atom()}
+        %% Connection is terminating, delay willmsg publish is impossible.
+        | ?chan_terminating
+        %% Connection is terminating because of normal MQTT disconnection, implies delete the session.
+        | normal.
+maybe_publish_will_msg(normal, Channel) ->
+    %% [MQTT-3.1.2-8]
     Channel;
-maybe_publish_will_msg(Channel = #channel{clientinfo = ClientInfo, will_msg = WillMsg}) ->
+maybe_publish_will_msg(_Reason, Channel = #channel{will_msg = undefined}) ->
+    %% No will message to publish
+    Channel;
+maybe_publish_will_msg(
+    _Reason,
+    Channel = #channel{
+        conn_state = ConnState,
+        conninfo = #{clientid := ClientId}
```

Comment:
Is it guaranteed to have a properly initialized conninfo?
the process may shutdown before gets assigned with a clientid.

e.g. CONNECT packet might be delayed processing due to being rate limited, and the process is to be shutdown for whatever reason.

Reply:
I think L2275 covers that?

Reply:
anyway, they are  init from here
https://github.com/emqx/emqx/blob/bb6a95bae809cbb13a13ce17307b018fc1729696/apps/emqx/src/emqx_channel.erl#L246

called from:
https://github.com/emqx/emqx/blob/bb6a95bae809cbb13a13ce17307b018fc1729696/apps/emqx/src/emqx_connection.erl#L349